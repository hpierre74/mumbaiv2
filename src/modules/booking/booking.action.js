import { setBooking, removeBooking } from '../../utils/booking.utils';

export const ADD_BOOKING = 'booking/ADD_BOOKING';
export const ADD_BOOKING_BEGIN = 'booking/ADD_BOOKING_BEGIN';
export const ADD_BOOKING_SUCCESS = 'booking/ADD_BOOKING_SUCCESS';
export const ADD_BOOKING_FAILURE = 'booking/ADD_BOOKING_FAILURE';

export const DELETE_BOOKING = 'booking/DELETE_BOOKING';
export const DELETE_BOOKING_BEGIN = 'booking/DELETE_BOOKING_BEGIN';
export const DELETE_BOOKING_SUCCESS = 'booking/DELETE_BOOKING_SUCCESS';
export const DELETE_BOOKING_FAILURE = 'booking/DELETE_BOOKING_FAILURE';

export const addBooking = booking => dispatch => {
  dispatch({ type: ADD_BOOKING_BEGIN });
  setBooking(booking)
    .then(success => dispatch({ type: ADD_BOOKING_SUCCESS, success }))
    .catch(err => dispatch({ type: ADD_BOOKING_FAILURE, err }));
};

export const deleteBooking = bookingId => dispatch => {
  dispatch({ type: DELETE_BOOKING_BEGIN });
  removeBooking(bookingId)
    .then(success => dispatch({ type: DELETE_BOOKING_SUCCESS, success }))
    .catch(err => dispatch({ type: DELETE_BOOKING_FAILURE, err }));
};
