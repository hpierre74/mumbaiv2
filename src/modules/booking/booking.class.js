import CryptoJS from 'crypto-js';
import isString from 'lodash/isString';
import isNumber from 'lodash/isNumber';
import isDate from 'lodash/isDate';

import { getData, setData, getNewKey, increment } from '../firebase/firebase.class';

export const getBookingsForDate = date =>
  getData(`booker/bookings`).then(_bookings =>
    Object.values(_bookings.val()).filter(_booking => _booking.date === date),
  );

export const getBookingsForService = bookings => service =>
  Object.values(bookings)
    .filter(booking => booking.service === service)
    .reduce((acc, _booking) => ({ ...acc, _booking }));

export const getPersonsForDateService = booking =>
  getData(`booker/bookings`).then(_bookings =>
    Object.values(_bookings.val())
      .filter(_booking => _booking.service === booking.service && _booking.date === booking.date)
      .map(_booking => _booking.persons)
      .reduce((acc, persons) => acc + persons),
  );

export const getPersonsForDate = booking =>
  getData(`booker/bookings`).then(_bookings =>
    Object.values(_bookings.val())
      .filter(_booking => _booking.date === booking.date)
      .map(_booking => _booking.persons)
      .reduce((acc, persons) => acc + persons),
  );

// export const getServiceForTime = (time, services) => {
//   Object.values(services).map(service => service.bookableHours.filter(hour => hour === time))

export const isBookingPossible = booking => maxBookings =>
  getPersonsForDateService(booking)
    .then(!(persons => persons + booking.persons > maxBookings))
    .catch(err => console.log(err));

export const addBooking = booking => {
  if (!isBookingPossible(booking)) {
    return false;
  }
  const bookingKey = getNewKey(`booker/bookings`);

  return setData(`bookings/${bookingKey}`, booking)
    .then(res => console.log(res))
    .catch(err => console.log(err));
};

export const getTwoFirstIntChar = time => parseInt(time.substring(0, 2), 10);

class Booking {
  constructor(booking, maxBookings) {
    this.booking = booking;
    this.maxBookings = maxBookings;
    this.dateObj = booking ? new Date(`${booking.date} ${booking.time}:00 GMT+0200`) : null;
  }

  isBookingPossible = () =>
    getPersonsForDateService(this.booking)
      .then(!(persons => persons + this.booking.persons > this.maxBookings))
      .catch(err => console.log(err));

  isBookingValid = () => {
    const { booking } = this;
    if (!isString(booking.firstname)) {
      return false;
    }
    if (!isString(booking.lastname)) {
      return false;
    }
    if (!isString(booking.tel)) {
      return false;
    }
    if (!isString(booking.comment)) {
      return false;
    }
    if (!isString(booking.email)) {
      return false;
    }
    if (!isNumber(booking.persons)) {
      return false;
    }

    return true;
  };

  get = id => {
    getData(`booker/bookings/${id}`).then(res => {
      const booking = res.val();
      this.booking = booking;
      this.dateObj = new Date(`${booking.date} ${booking.time}:00 GMT+0200`);
    });
  };

  getService = () =>
    getData('public/config/services').then(
      services =>
        Object.keys(services.val()).filter(
          serviceKey =>
            getTwoFirstIntChar(this.booking.time) >= getTwoFirstIntChar(services.val()[serviceKey].startTime) &&
            getTwoFirstIntChar(this.booking.time) <= getTwoFirstIntChar(services.val()[serviceKey].endTime),
        ),
      // Object.keys(services.val()).map(serviceKey => {
      //   const service = services.val()[serviceKey];

      //   const time = getTwoFirstIntChar(this.booking.time);

      //   if (time >= getTwoFirstIntChar(service.startTime) && time <= getTwoFirstIntChar(service.endTime)) {
      //     key = serviceKey;

      //     return key || null;
      //   }

      //   return key ? null : null;
      // }),
    );

  add = () => {
    // if (!this.isBookingValid()) {
    //   return;
    // }
    // if (!this.isBookingPossible()) {
    //   return;
    // }
    const bookingKey = getNewKey('booker/bookings');
    const updates = {};
    // const booking = CryptoJS.AES.encrypt(
    //   JSON.stringify({ ...this.booking, date: this.date, id: ref }),
    //   'secret key 123',
    // );
    this.getService(this.booking.time).then(key => {
      console.log(key);

      updates[`/bookings/${bookingKey}`] = {
        ...this.booking,
        id: bookingKey,
        dateObj: this.dateObj,
        service: key,
      };

      return setData(`booker`, updates).then(() => {
        increment('stats/bookings/total', 1);
        increment('stats/bookings/persons', this.booking.persons);
      });
    });
  };
  remove = () => {};
}
export default Booking;
