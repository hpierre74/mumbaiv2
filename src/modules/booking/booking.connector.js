import { connect } from 'react-redux';

import Booking from './booking.component';

const mapStateToProps = ({ app: { config } }) => ({
  config,
});

export default connect(mapStateToProps)(Booking);
