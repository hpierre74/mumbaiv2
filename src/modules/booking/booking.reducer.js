import {
  ADD_BOOKING_FAILURE,
  ADD_BOOKING_SUCCESS,
  DELETE_BOOKING_FAILURE,
  DELETE_BOOKING_SUCCESS,
} from './booking.action';

const initialState = {
  addBookingPending: false,
  addBookingError: null,
  addBookingSuccess: false,
  deleteBookingPending: false,
  deleteBookingError: null,
  deleteBookingSuccess: false,
};

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case ADD_BOOKING_SUCCESS:
      return {
        ...state,
        addBookingPending: false,
        addBookingSuccess: true,
      };

    case ADD_BOOKING_FAILURE:
      return {
        ...state,
        addBookingPending: false,
        addBookingError: action.error,
      };

    case DELETE_BOOKING_SUCCESS:
      // The request is success
      return {
        ...state,
        deleteBookingPending: false,
        deleteBookingSuccess: true,
      };

    case DELETE_BOOKING_FAILURE:
      // The request is failed
      return {
        ...state,
        deleteBookingPending: false,
        deleteBookingError: action.error,
      };

    default:
      return state;
  }
}
