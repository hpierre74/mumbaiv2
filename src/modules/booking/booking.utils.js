import { setData, getData, getNewKey } from '../../utils/firebase.utils';

export const getBooking = id => getData(`booker/bookings/${id}`);

export const getBookingByLastname = async lastname => {
  const bookings = await getData(`booker/bookings`);

  return Object.values(bookings).filter(booking => booking.lastname === lastname);
};

export const setBooking = booking => {
  const id = getNewKey('booker/bookings');
  const newBooking = { ...booking, id, date: booking.date.toDate() };

  return setData(`booker/bookings/${id}`, newBooking);
};

export const removeBooking = id => setData(`booker/bookings/${id}`, null);
