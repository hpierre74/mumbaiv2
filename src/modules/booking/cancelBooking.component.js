import React, { Component } from 'react';
import PropTypes from 'prop-types';
import omit from 'lodash/omit';

import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Card from '@material-ui/core/Card';
import CardHeader from '@material-ui/core/CardHeader';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';

import Confirm from '../../components/confirm.component';
import { Row, Col } from '../../components/grid.components';
import { requestPost } from '../../utils/http.utils';
// import { setData } from '../../utils/firebase.utils';

export default class CancelBooking extends Component {
  constructor(props) {
    super(props);
    this.defaultState = {
      bookingId: '',
      userEmail: '',
      modal: false,
    };
    this.state = { ...this.defaultState };
    this.toggleModal = this.toggleModal.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  toggleModal = () => {
    this.setState(state => ({ modal: !state.modal }));
  };

  handleSubmit = e => {
    e.preventDefault();
    this.toggleModal();
    const { showToast } = this.props;
    const state = omit(this.state, 'modal');
    console.log(state);

    requestPost('https://us-central1-la-taverne-du-perroquet-lyon.cloudfunctions.net/cancelBooking/', state)
      .then(res => {
        console.log(res);

        this.setState({ ...this.defaultState });
        showToast('success', 'Votre réservation a bien été annulée, merci');
      })
      .catch(err => showToast('error', "Oups, cela n'a pas marché", err));
  };

  render() {
    return (
      <Card style={{ margin: '2.5%' }}>
        <CardHeader color="secondary" component="h3" title="Annuler votre réservation" />
        <CardContent>
          <Row style={{ display: 'flex', justifyContent: 'center' }} container spacing={24}>
            <Col xs={12} md={6}>
              <TextField
                required
                onChange={this.handleChange}
                name="bookingId"
                value={this.state.bookingId}
                id="bookingId"
                label="Votre identifiant de réservation"
                fullWidth
              />
            </Col>
            <Col xs={12} md={6}>
              <TextField
                required
                onChange={this.handleChange}
                name="userEmail"
                value={this.state.userEmail}
                id="userEmail"
                label="Votre adresse email de réservation"
                fullWidth
              />
            </Col>
          </Row>
        </CardContent>
        <Row style={{ display: 'flex', justifyContent: 'center' }} container spacing={24}>
          <Col xs={12} md={12}>
            <Button onClick={this.toggleModal} variant="outlined" color="primary">
              Annuler la réservation
            </Button>
          </Col>
          <Col xs={12} md={12}>
            <Confirm
              title="ëtes vous sûr de vouloir annuler ?"
              open={this.state.modal}
              onCancel={this.toggleModal}
              onSubmit={this.handleSubmit}
            >
              {this.state.modal && (
                <Typography>{`Confirmer l'annulation de la réservation ${this.state.bookingId}`} </Typography>
              )}
            </Confirm>
          </Col>
        </Row>
      </Card>
    );
  }
}

CancelBooking.propTypes = {
  showToast: PropTypes.func.isRequired,
};
