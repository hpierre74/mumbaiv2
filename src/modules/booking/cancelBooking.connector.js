import { connect } from 'react-redux';

import CancelBooking from './cancelBooking.component';
import { showToast } from '../toaster/toaster.action';

const mapStateToProps = ({ toaster: { visible, content } }) => ({
  isToastVisible: visible,
  toastContent: content,
});

export default connect(
  mapStateToProps,
  { showToast },
)(CancelBooking);
