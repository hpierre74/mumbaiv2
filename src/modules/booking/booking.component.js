import React, { Component } from 'react';
import PropTypes from 'prop-types';

// import Recaptcha from 'react-recaptcha';

import { Title3 } from '../../components/title.components';
import { PageWrapper as BookingPageWrapper } from '../../components/wrapper.components';
import { Form, Input, Select, Textarea, Button } from '../../components/form.components';
import _Booking, { getBookingsForDate } from './booking.class';
import { setBooking, renderBookableHours } from '../../utils/booking.utils';
/* eslint-disable no-console */

class Booking extends Component {
  constructor(props) {
    super(props);
    this.state = {
      firstname: '',
      lastname: '',
      tel: '',
      email: '',
      persons: '',
      date: '',
      time: '',
      comment: '',
    };
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentDidMount = () => {};

  handleInputChange = e => {
    this.setState({
      [e.target.name]: e.target.value,
    });
  };

  handleSubmit = e => {
    e.preventDefault();
    const booking = new _Booking(this.state, this.props.config.maxBookings);
    booking.add();
    // setBooking(this.state);
    this.setState({
      firstname: '',
      lastname: '',
      tel: '',
      email: '',
      persons: 0,
      date: '',
      time: '',
      comment: '',
    });
  };

  render() {
    return (
      <BookingPageWrapper>
        <Title3>Prendre une Réservation</Title3>
        <Form onSubmit={this.handleSubmit}>
          <Input
            onChange={this.handleInputChange}
            value={this.state.firstname}
            name="firstname"
            placeholder="your firstname"
          />
          <Input
            onChange={this.handleInputChange}
            value={this.state.lastname}
            name="lastname"
            placeholder="your lastname"
          />
          <Input
            onChange={this.handleInputChange}
            value={this.state.tel}
            name="tel"
            placeholder="Votre numéro de réléphone"
          />
          <Input
            onChange={this.handleInputChange}
            value={this.state.email}
            name="email"
            placeholder="Votre adresse email"
          />
          <Input
            onChange={this.handleInputChange}
            value={this.state.persons}
            name="persons"
            placeholder="Le nombre de personnes attendues"
          />
          <Input
            type="date"
            onChange={this.handleInputChange}
            value={this.state.date}
            name="date"
            placeholder="Le jour de votre réservation"
          />
          <Select
            type="time"
            onChange={this.handleInputChange}
            value={this.state.time}
            name="time"
            placeholder="L'heure"
          >
            <option />
            {this.props.config !== null && renderBookableHours(this.props.config.services)}
          </Select>
          <Textarea
            onChange={this.handleInputChange}
            value={this.state.comment}
            name="comment"
            width="100%"
            placeholder="Allergies, bébés ..."
          />
          {/* <Recaptcha
            sitekey="6LcnsGAUAAAAAFPpq9NnMTVYgtdQU5q8zf4McL-W"
            render="implicit"
            onloadCallback={console.log('ok')}
            verifyCallback={console.log('verified')}
            theme="dark"
          /> */}
          <Button value="Envoyer" />
        </Form>
      </BookingPageWrapper>
    );
  }
}
Booking.defaultProps = {
  config: null,
};
Booking.propTypes = {
  config: PropTypes.shape({
    services: PropTypes.shape({}),
    maxBookings: PropTypes.number,
  }),
};

export default Booking;
