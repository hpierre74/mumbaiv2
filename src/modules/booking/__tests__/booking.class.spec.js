import * as firebase from '../../firebase/firebase.class';
import { shallow } from 'enzyme/';
import * as Booking from '../booking.class';

const bookings = {
  a: {
    persons: 5,
    firstname: 'a',
    lastname: 'a',
    date: '2018-09-25',
    time: '12:00',
    dateObj: new Date('2018-09-25 12:00:00 GMT+0200'),
    id: 'a',
    tel: '0000000000',
    email: 'a@a.fr',
    comment: '',
  },
  b: {
    persons: 5,
    firstname: 'a',
    lastname: 'b',
    date: '2018-09-25',
    time: '12:00',
    dateObj: new Date('2018-09-25 12:00:00 GMT+0200'),
    id: 'b',
    tel: '0000000000',
    email: 'a@a.fr',
    comment: '',
  },
  c: {
    persons: 5,
    firstname: 'a',
    lastname: 'c',
    date: '2018-09-25',
    time: '12:00',
    dateObj: new Date('2018-09-25 12:00:00 GMT+0200'),
    id: 'c',
    tel: '0000000000',
    email: 'a@a.fr',
    comment: '',
  },
  d: {
    persons: 5,
    firstname: 'a',
    lastname: 'd',
    date: '2018-09-25',
    time: '12:00',
    dateObj: new Date('2018-09-25 12:00:00 GMT+0200'),
    id: 'd',
    tel: '0000000000',
    email: 'a@a.fr',
    comment: '',
  },
};

firebase.getData = jest.fn(() => new Promise(res => res({ val: () => bookings })));

describe('getData mock', () =>
  it('return bookings', () => firebase.getData().then(response => expect(response.val()).toBe(bookings))));

describe('Booking Functions', () => {
  it('getBookingsForDate returns bookings', () =>
    Booking.getBookingsForDate('2018-09-25').then(res => expect(res).toMatchSnapshot()));
  it('getBookingsForDate returns nothing', () =>
    Booking.getBookingsForDate('2018-09-22').then(res => expect(res).toMatchSnapshot()));

  it('getPersonsForDateService returns persons', () =>
    Booking.getPersonsForDateService(bookings.a).then(res => expect(res).toBe(20)));
  it('getPersonsForDateService returns same persons', () =>
    Booking.getPersonsForDateService(bookings.b).then(res => expect(res).toBe(20)));
  it('getPersonsForDateService returns nothing', () =>
    Booking.getPersonsForDateService({ date: '2018-09-22', persons: 3 }).then(res => expect(res).toBe(3)));
});
