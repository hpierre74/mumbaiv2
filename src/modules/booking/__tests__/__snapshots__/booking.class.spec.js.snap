// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Booking Functions getBookingsForDate returns bookings 1`] = `
Array [
  Object {
    "comment": "",
    "date": "2018-09-25",
    "dateObj": 2018-09-25T10:00:00.000Z,
    "email": "a@a.fr",
    "firstname": "a",
    "id": "a",
    "lastname": "a",
    "persons": 5,
    "tel": "0000000000",
    "time": "12:00",
  },
  Object {
    "comment": "",
    "date": "2018-09-25",
    "dateObj": 2018-09-25T10:00:00.000Z,
    "email": "a@a.fr",
    "firstname": "a",
    "id": "b",
    "lastname": "b",
    "persons": 5,
    "tel": "0000000000",
    "time": "12:00",
  },
  Object {
    "comment": "",
    "date": "2018-09-25",
    "dateObj": 2018-09-25T10:00:00.000Z,
    "email": "a@a.fr",
    "firstname": "a",
    "id": "c",
    "lastname": "c",
    "persons": 5,
    "tel": "0000000000",
    "time": "12:00",
  },
  Object {
    "comment": "",
    "date": "2018-09-25",
    "dateObj": 2018-09-25T10:00:00.000Z,
    "email": "a@a.fr",
    "firstname": "a",
    "id": "d",
    "lastname": "d",
    "persons": 5,
    "tel": "0000000000",
    "time": "12:00",
  },
]
`;

exports[`Booking Functions getBookingsForDate returns nothing 1`] = `Array []`;
